### PR TITLE
Compare vendor with hack/versions and update hack/versions

### PR DESCRIPTION
**Pull request by miaoyq**
Part of #385

Compare vendor with hack/versions, update hack/versions if the commit id different from vendor.conf.
If OK, I will update `CONTRIBUTING.md`.

Signed-off-by: Yanqiang Miao <miao.yanqiang@zte.com.cn>

### CODE REVIEW

File: hack/update-vendor.sh [hunk 1]
```diff
@@ -18,15 +18,26 @@ set -o errexit
...
+map["RUNC_VERSION"]="github.com/opencontainers/runc"
+map["CNI_VERSION"]="github.com/containernetworking/cni"
+map["CONTAINERD_VERSION"]="github.com/containerd/containerd"
+map["KUBERNETES_VERSION"]="k8s.io/kubernetes"
+for key in ${!map[@]}
+do
+  vendor_commitid=$(grep ${map[${key}]} vendor.conf | awk '{print $2;}')
+  version_commitid=$(grep ${key} hack/versions | awk -F "=" '{print $2;}')
+  if [ ${vendor_commitid} != ${version_commitid} ]; then
+    echo "Update the value of ${key} from ${version_commitid} to ${vendor_commitid}."
```
Comment (mikebrow) [inline hack/update-vendor.sh]: /s/Update/Updating/

File: hack/update-vendor.sh [hunk 2]
```diff
@@ -18,15 +18,26 @@ set -o errexit
...
+echo "Compare vendor with hack/versions and update hack/versions..."
+declare -A map=()
+map["RUNC_VERSION"]="github.com/opencontainers/runc"
+map["CNI_VERSION"]="github.com/containernetworking/cni"
+map["CONTAINERD_VERSION"]="github.com/containerd/containerd"
+map["KUBERNETES_VERSION"]="k8s.io/kubernetes"
+for key in ${!map[@]}
+do
+  vendor_commitid=$(grep ${map[${key}]} vendor.conf | awk '{print $2;}')
+  version_commitid=$(grep ${key} hack/versions | awk -F "=" '{print $2;}')
```
Comment (mikebrow) [inline hack/update-vendor.sh]: are the semicolons necessary here?

**Comment by mikebrow** [comment]
I sort of prefer (2) above but (1) would also solve the issue.. wdyt?

**Comment by Random-Liu** [comment]
@mikebrow The problem with 2) is that cri-tools is not our dependency, it's not in vendor.conf.

**Comment by mikebrow** [comment]
@Random-Liu that's not really a problem in so far as at that point hack/versions would only have cri-tools in it and would only be specified in one location.

**Comment by Random-Liu** [comment]
> @Random-Liu that's not really a problem in so far as at that point hack/versions would only have cri-tools in it and would only be specified in one location.

In our document, we have some link to `hack/versions` to tell users what version of the dependencies we are using, which I think is also useful. :)

LGTM overall.
But as @mikebrow suggested, we should verify this in `make verify`.
You can add an option to the script `-w` or something, and format a temporary vendor.conf
* If `-w` is set, we should rewrite the vendor.conf file, just overwrite with the temporary vendor.conf.
* If `-w` is not set, we should only check whether temporary vendor.conf is the same with vendor.conf.

`-w` is just an example, you can use whichever flag or env you think is more convenient.

File: hack/update-vendor.sh (hunk 2 above)
Comment (miaoyq) [reply hack/update-vendor.sh]: No error, but I will remove it. :)

**Comment by yanxuean** [comment]
>In our document, we have some link to hack/versions to tell users what version of the dependencies we are using, which I think is also useful. :)

We may link the link to vndr.conf.  The document only state the dependency of cni, containerd and runc, not include cri-tool.

What do you think we add a function into hack/versions?  The function will get version from vndr.conf to set environment.  Of course it keep  cri-tools from modifying.
Because the hack/version also be a script, it only set some environment variable.

**Comment by Random-Liu** [comment]
@yanxuean `vendor.conf` has a lot of dependencies. :( It's hard to figure out which and which.

**Comment by yanxuean** [comment]
I think the user don't really concern the version of runC etc. They pay more attention to that if the cri-container  will work. 
For developer , we only want to modify it in one place.

**Comment by miaoyq** [comment]
>But as @mikebrow suggested, we should verify this in make verify.
You can add an option to the script -w or something, and format a temporary vendor.conf
>
>If -w is set, we should rewrite the vendor.conf file, just overwrite with the temporary vendor.conf.
If -w is not set, we should only check whether temporary vendor.conf is the same with vendor.conf.
-w is just an example, you can use whichever flag or env you think is more convenient.

I don't understand this clearly.
why do we need to format a temporary vendor.conf?
@Random-Liu

**Comment by Random-Liu** [comment]
@miaoyq We can work on the original vendor.conf, too. As long as we return non-zero exit when vendor is changed, it should be fine.

However, it feels to me that verification should not change the file. :)

**Comment by miaoyq** [comment]
@Random-Liu  This pr is updating the contents of `hack/versions` according to `vendor.conf`, 
Do you mean we should update `vendor.conf` according to `hack/versions`?

**Comment by miaoyq** [comment]
@Random-Liu I will add the verification to `make verify`.

**Comment by Random-Liu** [comment]
@miaoyq Sorry, I thought you were doing the other direction.

Yeah, we need to verify that in our CI so that we won't forget changing that in the future.

**Comment by miaoyq** [comment]
@Random-Liu Make senses, will do. :)

> @Random-Liu that's not really a problem in so far as at that point hack/versions would only have cri-tools in it and would only be specified in one location.

What @mikebrow  means is that `hack/versions` only contains the version of `cri-tools` and the versions of others is get from `vendor.conf` directly, right?  I think it is also a good solution, we can implement it later if necessary. :-)

File: hack/update-vendor.sh (hunk 2 above)
Comment (miaoyq) [reply hack/update-vendor.sh]: Done.

File: hack/update-vendor.sh (hunk 1 above)
Comment (miaoyq) [reply hack/update-vendor.sh]: Done.

**Comment by miaoyq** [comment]
```
$ hack/update-vendor.sh -only-verify
Sort vendor.conf...
Compare vendor with hack/versions...
Need to update the value of KUBERNETES_VERSION from 164317879bcd81eddws0b97e5ebf1c8df041770f2ff1b to 164317879bcd810b97e5ebf1c8df041770f2ff1b.
Please update "hack/versions" by executing "hack/update-vendor.sh"!
```

```
$ hack/update-vendor.sh                                   
Sort vendor.conf...
Compare vendor with hack/versions...
Updating the value of KUBERNETES_VERSION from 164317879bcfdsfdd810b97e5ebf1c8df041770f2ff1b to 164317879bcd810b97e5ebf1c8df041770f2ff1b.
Please commit the change made by this file...
```

**Comment by Random-Liu** [comment]
Nice!

**Pull request merged**